v1/plugins/logs/encoder.go

```diff
@@ -71,165 +91,376 @@ func (enc *chunkEncoder) WithMetrics(m metrics.Metrics) *chunkEncoder {
+		}
+
+		// The ND cache has to be dropped, record this size as a known maximum event size
+		if enc.maxEventSize == 0 || int64(len(eventBytes)) < enc.maxEventSize {
+			enc.maxEventSize = int64(len(eventBytes))
+		}
+
+		// 2. Drop the ND cache and see if the incoming event can fit within the current chunk without the cache (so we can maximize chunk size)
+		enc.initialize()
+		enc.incrMetric(encLogExUploadSizeLimitCounterName)
+		// If there's no ND builtins cache in the event, then we don't need to retry encoding anything.
+		if event.NDBuiltinCache == nil {
+			enc.incrMetric(logEncodingFailureCounterName)
+			if enc.logger != nil {
+				enc.logger.Error("Log encoding failed: received a decision event size (%d) that exceeded the upload_size_limit_bytes (%d). No ND cache to drop.", currentSize, enc.limit)
+			}
+			return nil, nil
+		}
+		// re-encode the event with the ND cache removed
+		event.NDBuiltinCache = nil
+
+		eventBytes, err = json.Marshal(&event)
 		if err != nil {
 			return nil, err
 		}
-		enc.bytesWritten += n
+		err = enc.appendEvent(eventBytes)
+		if err != nil {
+			return nil, err
+		}
+
+		result, err = enc.reset()
+		if err != nil {
+			return nil, err
+		}
+
+		if len(result[0]) > int(enc.limit) {
+			enc.incrMetric(logEncodingFailureCounterName)
+			if enc.logger != nil {
+				enc.logger.Error("Log encoding failed: received a decision event size (%d) that exceeded the upload_size_limit_bytes (%d) even after dropping the ND cache.",
+					len(eventBytes), enc.limit)
+			}
+			enc.initialize() // drop the event
+			return nil, nil
+		}
+
+		// success! the incoming event without its ND cache fits into the current chunk
+		err = enc.appendEvent(eventBytes)
+		if err != nil {
+			return nil, err
+		}
+		if enc.logger != nil {
+			enc.logger.Error("ND builtins cache dropped from this event to fit under maximum upload size limits. Increase upload size limit or change usage of non-deterministic builtins.")
+		}
+		enc.incrMetric(logNDBDropCounterName)
+		// success! the incoming chunk lost the ND cache, but it wasn't dropped entirely
+		// scale up the uncompressed limit using the uncompressed event size as a base
+		if int64(len(eventBytes)) > enc.uncompressedLimit {
+			enc.uncompressedLimit = int64(len(eventBytes))
+		}
+		enc.scaleUp()
+		return nil, nil
 	}
 
-	n, err := enc.w.Write(bs)
+	enc.updateMetric(encNumberOfEventsInChunkHistogramName, enc.eventsWritten)
+	result, err := enc.reset()
 	if err != nil {
 		return nil, err
 	}
 
-	enc.bytesWritten += n
-	return
-}
+	// 1) Scale Up: If the current chunk size is within 90% of the user-configured limit, exponentially increase
+	// the uncompressed limit. The exponential function is 2^x where x has a minimum value of 1
+	if len(result[0]) < enc.threshold {
+		enc.scaleUp()
 
-func (enc *chunkEncoder) writeClose() error {
-	if _, err := enc.w.Write([]byte(`]`)); err != nil {
-		return err
+		r, err := enc.Encode(event, eventBytes)
+		if err != nil {
+			return nil, err
```

Comment:
If the new event is too large, and can't be added to the encoder, we're dropping the `result` here. Don't we know that it contains a chunk by this point, which we're now dropping despite nothing being wrong with the contained events?

Reply:
Good catch, returning `result` is the right thing to do